packages/@ember/-internals/glimmer/lib/modifiers/on.ts

```diff
@@ -0,0 +1,241 @@
+): void {
+  removes++;
+
+  if (SUPPORTS_EVENT_OPTIONS) {
+    // when options are supported, use them across the board
+    element.removeEventListener(eventName, callback, options);
+  } else if (options !== undefined && options.capture) {
+    // used only in the following case:
+    //
+    // `{ once: true | false, passive: true | false, capture: true }
+    //
+    // `once` is handled via a custom callback that removes after first
+    // invocation so we only care about capture here as a boolean
+    element.removeEventListener(eventName, callback, true);
+  } else {
+    // used only in the following cases:
+    //
+    // * where there is no options
+    // * `{ once: true | false, passive: true, capture: false }
+    element.removeEventListener(eventName, callback);
+  }
+}
+
+function addEventListener(
+  element: Element,
+  eventName: string,
+  callback: EventListener,
+  options?: AddEventListenerOptions
+): void {
+  adds++;
+
+  if (SUPPORTS_EVENT_OPTIONS) {
+    // when options are supported, use them across the board
+    element.addEventListener(eventName, callback, options);
+  } else if (options !== undefined && options.capture) {
+    // used only in the following case:
+    //
+    // `{ once: true | false, passive: true | false, capture: true }
+    //
+    // `once` is handled via a custom callback that removes after first
+    // invocation so we only care about capture here as a boolean
+    element.addEventListener(eventName, callback, true);
+  } else {
+    // used only in the following cases:
+    //
+    // * where there is no options
+    // * `{ once: true | false, passive: true, capture: false }
+    element.addEventListener(eventName, callback);
+  }
+}
+
+export default class OnModifierManager implements ModifierManager<OnModifierState, Opaque> {
+  public SUPPORTS_EVENT_OPTIONS: boolean = SUPPORTS_EVENT_OPTIONS;
+
+  get counters() {
+    return { adds, removes };
+  }
+
+  create(element: Simple.Element | Element, _state: Opaque, args: Arguments) {
+    const capturedArgs = args.capture();
+
+    return new OnModifierState(<Element>element, capturedArgs);
+  }
+
+  getTag({ tag }: OnModifierState): Tag {
+    return tag;
+  }
+
+  install(state: OnModifierState) {
+    state.updateFromArgs();
+
+    let { element, eventName, callback, options } = state;
+
+    addEventListener(element, eventName, callback, options);
+
+    state.shouldUpdate = false;
+  }
+
+  update(state: OnModifierState) {
+    // stash prior state for el.removeEventListener
+    let { element, eventName, callback, options } = state;
+
+    state.updateFromArgs();
+
+    if (!state.shouldUpdate) {
+      return;
+    }
+
+    // use prior state values for removal
+    removeEventListener(element, eventName, callback, options);
+
+    // read updated values from the state object
+    addEventListener(state.element, state.eventName, state.callback, state.options);
+
+    state.shouldUpdate = false;
+  }
```

Comment:
If I'm reading this correctly, once the modifier is installed, it will only ever update to a new listener, but will never be uninstalled. I can't think of a scenario, where the modifier would need to be uninstalled from an element that _does not_ get removed from the DOM afterwards, but I think @chancancode made a case for this once on Discord.

Reply:
👉 https://twitter.com/chancancode/status/1109502949972074496

Reply:
@buschtoens - We definitely cleanup (see `getDestructor` just below, and the `destroy` method on the `OnModifierState` object. I will also add a test (using the assert count methodology) to ensure we don't accidentally regress this.

Reply:
Ah, I see! So `getDestructor` returns a `Destroyable`, which is implemented through `destroy()` in `OnModifierState`. I was not familiar with that part of the Glimmer API yet. 👍

Reply:
Ya, this is one way the "internal" modifiers differ from the public API. The public API _effectively_ does the same thing (the `destroy` method on the state bucket invokes the `destroyModifier` hook on the custom manager).